File: problems/problems_552/Solution.java
package problems.problems_552;

import com.alibaba.fastjson.JSON;
import java.util.*;
import qubhjava.BaseSolution;


public class Solution extends BaseSolution {
    int N = 6;
    int mod = (int)1e9+7;
    long[][] mul(long[][] a, long[][] b) {
        int r = a.length, c = b[0].length, z = b.length;
        long[][] ans = new long[r][c];
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                for (int k = 0; k < z; k++) {
                    ans[i][j] += a[i][k] * b[k][j];
                    ans[i][j] %= mod;
                }
            }
        }
        return ans;
    }
    public int checkRecord(int n) {
        long[][] ans = new long[][]{
            {1}, {0}, {0}, {0}, {0}, {0}
        };
        long[][] mat = new long[][]{
            {1, 1, 1, 0, 0, 0},
            {1, 0, 0, 0, 0, 0},
            {0, 1, 0, 0, 0, 0},
            {1, 1, 1, 1, 1, 1},
            {0, 0, 0, 1, 0, 0},
            {0, 0, 0, 0, 1, 0}
        };
        while (n != 0) {
            if ((n & 1) != 0) ans = mul(mat, ans);
            mat = mul(mat, mat);
            n >>= 1;
        }
        int res = 0;
        for (int i = 0; i < N; i++) {
            res += ans[i][0];
            res %= mod;
        }
        return res;
    }

    @Override
    public Object solve(String[] inputJsonValues) {
        int n = Integer.parseInt(inputJsonValues[0]);
        return JSON.toJSON(checkRecord(n));
    }
}

File: problems/problems_552/Solution.cpp
//go:build ignore
#include "cpp/common/Solution.h"

using namespace std;
using json = nlohmann::json;

class Solution {
  static constexpr int MOD = 1'000'000'007;
  static constexpr int SIZE = 6;

  using matrix = array<array<int, SIZE>, SIZE>;

  // 返回矩阵 a 和矩阵 b 相乘的结果
  matrix mul(matrix &a, matrix &b) {
    matrix c{};
    for (int i = 0; i < SIZE; i++) {
      for (int j = 0; j < SIZE; j++) {
        for (int k = 0; k < SIZE; k++) {
          c[i][j] = (c[i][j] + (long long)a[i][k] * b[k][j]) % MOD;
        }
      }
    }
    return c;
  }

  // 返回 n 个矩阵 a 相乘的结果
  matrix pow(matrix a, int n) {
    matrix res = {};
    for (int i = 0; i < SIZE; i++) {
      res[i][i] = 1; // 单位矩阵
    }
    while (n) {
      if (n & 1) {
        res = mul(res, a);
      }
      a = mul(a, a);
      n >>= 1;
    }
    return res;
  }

public:
  int checkRecord(int n) {
    matrix m = {{
        {{1, 1, 0, 1, 0, 0}},
        {{1, 0, 1, 1, 0, 0}},
        {{1, 0, 0, 1, 0, 0}},
        {{0, 0, 0, 1, 1, 0}},
        {{0, 0, 0, 1, 0, 1}},
        {{0, 0, 0, 1, 0, 0}},
    }};
    matrix res = pow(m, n);
    int ans = 0;
    for (int x : res[0]) {
      ans = (ans + x) % MOD;
    }
    return ans;
  }
};

json leetcode::qubh::Solve(string input_json_values) {
  vector<string> inputArray;
  size_t pos = input_json_values.find('\n');
  while (pos != string::npos) {
    inputArray.push_back(input_json_values.substr(0, pos));
    input_json_values = input_json_values.substr(pos + 1);
    pos = input_json_values.find('\n');
  }
  inputArray.push_back(input_json_values);

  Solution solution;
  int n = json::parse(inputArray.at(0));
  return solution.checkRecord(n);
}
